File: lib/keychain_access.dart

import 'keychain_access_platform_interface.dart';

class KeychainAccess {

  /// Add password with key and value(password).
  /// Optionally, specify application.
  Future<bool> addPassword(String key, String value, {
    String? application
  }) {
    return KeychainAccessPlatform.instance.addPassword(
        key,
        value,
        application: application
    );
  }

  /// Update password with key and value(password).
  /// Optionally, specify application.
  Future<bool> updatePassword(String key, String value, {
    String? application
  }) {
    return KeychainAccessPlatform.instance.updatePassword(
        key,
        value,
        application: application
    );
  }

  /// addOrUpdate password with key and value(password).
  /// Optionally, specify application.
  Future<bool> addOrUpdatePassword(String key, String value, {
    String? application
  }) {
    return KeychainAccessPlatform.instance.addOrUpdatePassword(
        key,
        value,
        application: application
    );
  }

  /// Find password for key.
  /// Optionally, specify application.
  Future<String?> findPassword(String key, {
    String? application
  }) {
    return KeychainAccessPlatform.instance.findPassword(
        key,
        application: application
    );
  }

  /// Delete password for key.
  /// Optionally, specify application.
  Future<bool> deletePassword(String key, {
    String? application
  }) {
    return KeychainAccessPlatform.instance.deletePassword(
        key,
        application: application
    );
  }
}

File: macos/Classes/KeychainAccess.swift
//
//  KeychainAccess.swift
//  keychain_access
//
//  Created by Paul Gundarapu on 4/4/24.
//

import Foundation
import Security

class KeychainAccess {
    // Function to add a password to the keychain
    static func addPassword(service:String?, account: String, password: String) -> OSStatus {
        let query = NSMutableDictionary()
        query[kSecClass] = kSecClassGenericPassword
        query[kSecAttrAccount] = account
        query[kSecValueData] = Data(password.utf8)
        if service != nil {
            query[kSecAttrService] = service
        }
        
        return SecItemAdd(query, nil)
    }
    
    
    // Function to update the password to the keychain
    static func updatePassword(service:String?, account: String, password: String) -> OSStatus {
        let query = NSMutableDictionary()
        query[kSecClass] = kSecClassGenericPassword
        query[kSecAttrAccount] = account
        if service != nil {
            query[kSecAttrService] = service
        }
        
        let attributesToUpdate = [
            kSecValueData: Data(password.utf8)
        ] as CFDictionary
        
        return SecItemUpdate(query, attributesToUpdate)
    }

    // Function to find a password in the keychain
    static func findPassword(service: String?, account: String) -> (OSStatus, String?) {
        let query = NSMutableDictionary()
        query[kSecClass] = kSecClassGenericPassword
        query[kSecAttrAccount] = account
        query[kSecMatchLimit] = kSecMatchLimitOne
        query[kSecReturnData] = kCFBooleanTrue as Any
        
        if service != nil {
            query[kSecAttrService] = service
        }
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query, &result)
        
        guard status == noErr, let data = result as? Data else {
            return (status, nil)
        }
        
        return (status, String(data: data, encoding: .utf8))
    }

    // Function to delete a password from the keychain
    static func deletePassword(service: String?, account: String) -> OSStatus {
        let query = NSMutableDictionary()
        query[kSecClass] = kSecClassGenericPassword
        query[kSecAttrAccount] = account
        
        if service != nil {
            query[kSecAttrService] = service
        }
        
        return SecItemDelete(query)
    }
}
